src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+    private Set<LazyAttribute> initializedAttributes = new HashSet<LazyAttribute>();
+
+    private enum LazyField {
+        ALIGNMENT_START,
+        MAPPING_QUALITY,
+        REFERENCE_NAME,
+        CIGAR_STRING,
+        BASES,
+        QUALS,
+        MATE_ALIGNMENT_START,
+        MATE_REFERENCE_NAME,
+        INFERRED_INSERT_SIZE
+    }
+
+    private enum LazyFlag {
+        READ_NEGATIVE_STRAND,
+        READ_PAIRED,
+        NOT_PRIMARY_ALIGNMENT,
+        MATE_NEGATIVE_STRAND,
+        MATE_UNMAPPED,
+        FIRST_OF_PAIR,
+        SECOND_OF_PAIR
+    }
+
+    private enum LazyAttribute {
+        RG
+    }
+
+    private static class FlagGetter {
+        String getterName;
+        boolean canCallOnNotPaired;
+
+        FlagGetter(String getterName, boolean canCallOnAligned) {
+            this.getterName = getterName;
+            this.canCallOnNotPaired = canCallOnAligned;
+        }
+    }
+
+    private static class AttributeTagInfo {
+        LazyAttribute lazyAttribute;
+        String getterImplName;
+
+        AttributeTagInfo(LazyAttribute lazyAttribute, String getterImplName) {
+            this.lazyAttribute = lazyAttribute;
+            this.getterImplName = getterImplName;
+        }
+    }
+
+    private static Map<LazyFlag, FlagGetter> lazyFlagsGetters;
+    static
+    {
+        lazyFlagsGetters = new HashMap<LazyFlag, FlagGetter>();
+        lazyFlagsGetters.put(LazyFlag.READ_NEGATIVE_STRAND, new FlagGetter("getReadNegativeStrandFlag", true));
+        lazyFlagsGetters.put(LazyFlag.READ_PAIRED, new FlagGetter("getReadPairedFlag", true));
+        lazyFlagsGetters.put(LazyFlag.NOT_PRIMARY_ALIGNMENT, new FlagGetter("getNotPrimaryAlignmentFlag", true));
+        lazyFlagsGetters.put(LazyFlag.MATE_NEGATIVE_STRAND, new FlagGetter("getMateNegativeStrandFlag", false));
+        lazyFlagsGetters.put(LazyFlag.MATE_UNMAPPED, new FlagGetter("getMateUnmappedFlag", false));
+        lazyFlagsGetters.put(LazyFlag.FIRST_OF_PAIR, new FlagGetter("getFirstOfPairFlag", false));
+        lazyFlagsGetters.put(LazyFlag.SECOND_OF_PAIR, new FlagGetter("getSecondOfPairFlag", false));
+    }
+
+    private static Map<Short, AttributeTagInfo> lazyAttributeTags;
+    static
+    {
+        lazyAttributeTags = new HashMap<Short, AttributeTagInfo>();
+        lazyAttributeTags.put(SAMTagUtil.getSingleton().RG, new AttributeTagInfo(LazyAttribute.RG, "getAttributeGroupNameImpl"));
+    }
+
+    public SRALazyRecord(final SAMFileHeader header, boolean isAligned, ReadCollection run, Object sraIterator, String readId, String alignmentId) {
+        super(header);
+
+        if (!isAligned) {
+            throw new RuntimeException("Wrong constructor called. Call constructor which accepts unalignedReadFragmentIndex");
+        }
+        if (!(sraIterator instanceof AlignmentIterator)) {
+            throw new RuntimeException("Wrong SRA iterator provided. Expected AlignmentIterator");
+        }
```

Comment:
This is awkward -- why not have the constructors accept strongly-typed iterators  (ie., parameters of type `AlignmentIterator` and `ReadIterator` instead of `Object`), and get rid of the `isAligned` parameter completely, since you can infer its value from the type of iterator passed in?